File: tests/repos/go-calculator/operations/advanced.go
package operations

import (
	"context"
	"errors"
	"fmt"
	"math"
	"sync"
	"time"
)

// OperationType represents the type of operation.
type OperationType int

const (
	// OpUnary represents a unary operation.
	OpUnary OperationType = iota
	// OpBinary represents a binary operation.
	OpBinary
	// OpReduction represents a reduction operation.
	OpReduction
)

// CalculationError represents an error during calculation.
type CalculationError struct {
	Operation string
	Message   string
	Err       error
}

func (e *CalculationError) Error() string {
	if e.Err != nil {
		return fmt.Sprintf("%s: %s: %v", e.Operation, e.Message, e.Err)
	}
	return fmt.Sprintf("%s: %s", e.Operation, e.Message)
}

func (e *CalculationError) Unwrap() error {
	return e.Err
}

// OperationResult holds the result of a calculation.
type OperationResult struct {
	Value     float64
	Operation string
	Inputs    []float64
	Success   bool
	Error     error
	Duration  time.Duration
}

// HistoryEntry represents an entry in calculation history.
type HistoryEntry struct {
	Expression string
	Result     float64
	Timestamp  time.Time
}

// Calculator defines the interface for calculators.
type Calculator interface {
	Calculate(op string, args ...float64) (*OperationResult, error)
	Reset()
	GetHistory() []HistoryEntry
}

// AdvancedCalculator provides advanced calculation capabilities.
type AdvancedCalculator struct {
	precision    int
	memory       float64
	history      []HistoryEntry
	historyLimit int
	mu           sync.RWMutex
}

// NewAdvancedCalculator creates a new AdvancedCalculator instance.
func NewAdvancedCalculator(precision, historyLimit int) *AdvancedCalculator {
	return &AdvancedCalculator{
		precision:    precision,
		memory:       0,
		history:      make([]HistoryEntry, 0, historyLimit),
		historyLimit: historyLimit,
	}
}

// Calculate performs the specified operation.
func (c *AdvancedCalculator) Calculate(op string, args ...float64) (*OperationResult, error) {
	start := time.Now()

	c.mu.Lock()
	defer c.mu.Unlock()

	result := &OperationResult{
		Operation: op,
		Inputs:    args,
	}

	var value float64
	var err error

	// Binary operations
	switch op {
	case "+":
		if len(args) != 2 {
			err = errors.New("addition requires 2 arguments")
		} else {
			value = Add(args[0], args[1])
		}
	case "-":
		if len(args) != 2 {
			err = errors.New("subtraction requires 2 arguments")
		} else {
			value = Subtract(args[0], args[1])
		}
	case "*":
		if len(args) != 2 {
			err = errors.New("multiplication requires 2 arguments")
		} else {
			value = Multiply(args[0], args[1])
		}
	case "/":
		if len(args) != 2 {
			err = errors.New("division requires 2 arguments")
		} else {
			value, err = Divide(args[0], args[1])
		}
	case "**", "pow":
		if len(args) != 2 {
			err = errors.New("power requires 2 arguments")
		} else {
			value = Power(args[0], args[1])
		}
	case "%", "mod":
		if len(args) != 2 {
			err = errors.New("modulo requires 2 arguments")
		} else {
			intResult, modErr := Modulo(int(args[0]), int(args[1]))
			if modErr != nil {
				err = modErr
			} else {
				value = float64(intResult)
			}
		}

	// Unary operations
	case "sqrt":
		if len(args) != 1 {
			err = errors.New("sqrt requires 1 argument")
		} else if args[0] < 0 {
			err = &CalculationError{
				Operation: "sqrt",
				Message:   "cannot take square root of negative number",
			}
		} else {
			value = math.Sqrt(args[0])
		}
	case "log":
		if len(args) < 1 || len(args) > 2 {
			err = errors.New("log requires 1 or 2 arguments")
		} else if args[0] <= 0 {
			err = &CalculationError{
				Operation: "log",
				Message:   "cannot take log of non-positive number",
			}
		} else {
			if len(args) == 2 {
				value = math.Log(args[0]) / math.Log(args[1])
			} else {
				value = math.Log(args[0])
			}
		}
	case "sin":
		if len(args) != 1 {
			err = errors.New("sin requires 1 argument")
		} else {
			value = math.Sin(args[0])
		}
	case "cos":
		if len(args) != 1 {
			err = errors.New("cos requires 1 argument")
		} else {
			value = math.Cos(args[0])
		}
	case "tan":
		if len(args) != 1 {
			err = errors.New("tan requires 1 argument")
		} else {
			value = math.Tan(args[0])
		}
	case "abs":
		if len(args) != 1 {
			err = errors.New("abs requires 1 argument")
		} else {
			value = math.Abs(args[0])
		}
	case "floor":
		if len(args) != 1 {
			err = errors.New("floor requires 1 argument")
		} else {
			value = math.Floor(args[0])
		}
	case "ceil":
		if len(args) != 1 {
			err = errors.New("ceil requires 1 argument")
		} else {
			value = math.Ceil(args[0])
		}

	// Reduction operations
	case "sum":
		value = Sum(args...)
	case "product":
		value = Product(args...)
	case "max":
		if len(args) == 0 {
			err = errors.New("max requires at least 1 argument")
		} else {
			value = args[0]
			for _, a := range args[1:] {
				if a > value {
					value = a
				}
			}
		}
	case "min":
		if len(args) == 0 {
			err = errors.New("min requires at least 1 argument")
		} else {
			value = args[0]
			for _, a := range args[1:] {
				if a < value {
					value = a
				}
			}
		}
	case "avg", "mean":
		if len(args) == 0 {
			err = errors.New("average requires at least 1 argument")
		} else {
			value = Sum(args...) / float64(len(args))
		}

	default:
		err = &CalculationError{
			Operation: op,
			Message:   "unknown operation",
		}
	}

	result.Duration = time.Since(start)

	if err != nil {
		result.Success = false
		result.Error = err
		return result, err
	}

	// Round to precision
	multiplier := math.Pow(10, float64(c.precision))
	value = math.Round(value*multiplier) / multiplier

	result.Value = value
	result.Success = true

	// Add to history
	c.addHistory(op, args, value)

	return result, nil
}

// addHistory adds an entry to the calculation history.
func (c *AdvancedCalculator) addHistory(op string, args []float64, result float64) {
	expr := fmt.Sprintf("%s(%v)", op, args)
	entry := HistoryEntry{
		Expression: expr,
		Result:     result,
		Timestamp:  time.Now(),
	}

	if len(c.history) >= c.historyLimit {
		c.history = c.history[1:]
	}
	c.history = append(c.history, entry)
}

// Reset clears the calculator state.
func (c *AdvancedCalculator) Reset() {
	c.mu.Lock()
	defer c.mu.Unlock()

	c.memory = 0
	c.history = c.history[:0]
}

// GetHistory returns the calculation history.
func (c *AdvancedCalculator) GetHistory() []HistoryEntry {
	c.mu.RLock()
	defer c.mu.RUnlock()

	result := make([]HistoryEntry, len(c.history))
	copy(result, c.history)
	return result
}

// MemoryAdd adds a value to memory.
func (c *AdvancedCalculator) MemoryAdd(value float64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.memory += value
}

// MemorySubtract subtracts a value from memory.
func (c *AdvancedCalculator) MemorySubtract(value float64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.memory -= value
}

// MemoryClear clears the memory.
func (c *AdvancedCalculator) MemoryClear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.memory = 0
}

// MemoryRecall returns the memory value.
func (c *AdvancedCalculator) MemoryRecall() float64 {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.memory
}

// ScientificCalculator extends AdvancedCalculator with scientific functions.
type ScientificCalculator struct {
	*AdvancedCalculator
	angleMode string // "radians" or "degrees"
}

// NewScientificCalculator creates a new ScientificCalculator.
func NewScientificCalculator(precision, historyLimit int) *ScientificCalculator {
	return &ScientificCalculator{
		AdvancedCalculator: NewAdvancedCalculator(precision, historyLimit),
		angleMode:          "radians",
	}
}

// SetAngleMode sets the angle mode for trigonometric functions.
func (c *ScientificCalculator) SetAngleMode(mode string) error {
	if mode != "radians" && mode != "degrees" {
		return errors.New("angle mode must be 'radians' or 'degrees'")
	}
	c.angleMode = mode
	return nil
}

// ToRadians converts an angle to radians if in degrees mode.
func (c *ScientificCalculator) ToRadians(angle float64) float64 {
	if c.angleMode == "degrees" {
		return angle * math.Pi / 180
	}
	return angle
}

// Factorial calculates n! iteratively.
func (c *ScientificCalculator) Factorial(n int) (int64, error) {
	if n < 0 {
		return 0, &CalculationError{
			Operation: "factorial",
			Message:   "negative input not allowed",
		}
	}
	if n > 20 {
		return 0, &CalculationError{
			Operation: "factorial",
			Message:   "input too large (max 20)",
		}
	}

	var result int64 = 1
	for i := 2; i <= n; i++ {
		result *= int64(i)
	}
	return result, nil
}

// Fibonacci calculates the nth Fibonacci number.
func (c *ScientificCalculator) Fibonacci(n int) (int64, error) {
	if n < 0 {
		return 0, &CalculationError{
			Operation: "fibonacci",
			Message:   "negative index not allowed",
		}
	}

	if n <= 1 {
		return int64(n), nil
	}

	var a, b int64 = 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b, nil
}

// IsPrime checks if a number is prime.
func (c *ScientificCalculator) IsPrime(n int) bool {
	if n < 2 {
		return false
	}
	if n == 2 {
		return true
	}
	if n%2 == 0 {
		return false
	}

	sqrt := int(math.Sqrt(float64(n)))
	for i := 3; i <= sqrt; i += 2 {
		if n%i == 0 {
			return false
		}
	}
	return true
}

// PrimeFactors returns the prime factors of n.
func (c *ScientificCalculator) PrimeFactors(n int) []int {
	var factors []int
	d := 2
	for d*d <= n {
		for n%d == 0 {
			factors = append(factors, d)
			n /= d
		}
		d++
	}
	if n > 1 {
		factors = append(factors, n)
	}
	return factors
}

// GCD calculates the greatest common divisor using Euclidean algorithm.
func (c *ScientificCalculator) GCD(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return Abs(a)
}

// LCM calculates the least common multiple.
func (c *ScientificCalculator) LCM(a, b int) int {
	if a == 0 || b == 0 {
		return 0
	}
	return Abs(a*b) / c.GCD(a, b)
}

// CalculateWithContext performs calculation with context support for cancellation.
func (c *ScientificCalculator) CalculateWithContext(
	ctx context.Context,
	op string,
	args ...float64,
) (*OperationResult, error) {
	// Check if context is already cancelled
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	default:
	}

	// Perform calculation
	result, err := c.Calculate(op, args...)

	// Check again after calculation
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	default:
	}

	return result, err
}

// BatchCalculate performs multiple calculations concurrently.
func (c *ScientificCalculator) BatchCalculate(
	ctx context.Context,
	operations []struct {
		Op   string
		Args []float64
	},
) []*OperationResult {
	results := make([]*OperationResult, len(operations))
	var wg sync.WaitGroup

	for i, op := range operations {
		wg.Add(1)
		go func(idx int, operation struct {
			Op   string
			Args []float64
		}) {
			defer wg.Done()

			select {
			case <-ctx.Done():
				results[idx] = &OperationResult{
					Operation: operation.Op,
					Success:   false,
					Error:     ctx.Err(),
				}
				return
			default:
			}

			result, _ := c.Calculate(operation.Op, operation.Args...)
			results[idx] = result
		}(i, op)
	}

	wg.Wait()
	return results
}

File: tests/repos/typescript-calculator/src/operations/advanced.ts
/**
 * Advanced calculator with class-based design and various OOP patterns.
 */

// Import types
import type {
    CalculatorConfig,
    OperationResult,
    HistoryEntry,
    Observable,
    Observer,
    Unsubscribe,
    AngleMode,
} from './types.js';

// Import values
import {
    CalculationError,
    ErrorCode,
    DEFAULT_CONFIG,
} from './types.js';

// Import from sibling module
import { add, subtract, multiply, divide, sum, product } from './basic.js';

/**
 * Abstract base class for calculators.
 */
export abstract class BaseCalculator {
    protected _precision: number;
    protected _history: HistoryEntry[] = [];

    constructor(config: Partial<CalculatorConfig> = {}) {
        this._precision = config.precision ?? DEFAULT_CONFIG.precision;
    }

    // Abstract methods
    abstract calculate(op: string, ...args: number[]): OperationResult;
    abstract reset(): void;

    // Protected method
    protected roundResult(value: number): number {
        const multiplier = Math.pow(10, this._precision);
        return Math.round(value * multiplier) / multiplier;
    }

    // Getter
    get precision(): number {
        return this._precision;
    }

    // Setter with validation
    set precision(value: number) {
        if (value < 0 || !Number.isInteger(value)) {
            throw new Error('Precision must be a non-negative integer');
        }
        this._precision = value;
    }

    get history(): HistoryEntry[] {
        return [...this._history];
    }
}

/**
 * Advanced calculator implementation.
 */
export class AdvancedCalculator extends BaseCalculator implements Observable<OperationResult> {
    private static instanceCount = 0;

    private _memory: number = 0;
    private _historyLimit: number;
    private _lastResult: number = 0;
    private _observers: Observer<OperationResult>[] = [];

    constructor(config: Partial<CalculatorConfig> = {}) {
        super(config);
        this._historyLimit = config.historyLimit ?? DEFAULT_CONFIG.historyLimit;
        AdvancedCalculator.instanceCount++;
    }

    // Static method
    static getInstanceCount(): number {
        return AdvancedCalculator.instanceCount;
    }

    // Static factory method
    static create(config?: Partial<CalculatorConfig>): AdvancedCalculator {
        return new AdvancedCalculator(config);
    }

    // Implement abstract method
    calculate(op: string, ...args: number[]): OperationResult {
        const startTime = performance.now();

        try {
            let value: number;

            // Switch statement for operation dispatch
            switch (op) {
                case '+':
                case 'add':
                    if (args.length !== 2) throw new Error('Add requires 2 arguments');
                    value = add(args[0], args[1]);
                    break;

                case '-':
                case 'subtract':
                    if (args.length !== 2) throw new Error('Subtract requires 2 arguments');
                    value = subtract(args[0], args[1]);
                    break;

                case '*':
                case 'multiply':
                    if (args.length !== 2) throw new Error('Multiply requires 2 arguments');
                    value = multiply(args[0], args[1]);
                    break;

                case '/':
                case 'divide': {
                    if (args.length !== 2) throw new Error('Divide requires 2 arguments');
                    const result = divide(args[0], args[1]);
                    if (result.isErr()) throw result.error;
                    value = result.unwrap();
                    break;
                }

                case '**':
                case 'pow':
                case 'power':
                    if (args.length !== 2) throw new Error('Power requires 2 arguments');
                    value = Math.pow(args[0], args[1]);
                    break;

                case '%':
                case 'mod':
                case 'modulo':
                    if (args.length !== 2) throw new Error('Modulo requires 2 arguments');
                    if (args[1] === 0) throw new CalculationError('Modulo by zero', ErrorCode.DivisionByZero);
                    value = args[0] % args[1];
                    break;

                // Unary operations
                case 'sqrt':
                    if (args.length !== 1) throw new Error('Sqrt requires 1 argument');
                    if (args[0] < 0) throw new CalculationError('Cannot sqrt negative', ErrorCode.NegativeInput);
                    value = Math.sqrt(args[0]);
                    break;

                case 'log':
                    if (args.length < 1 || args.length > 2) throw new Error('Log requires 1-2 arguments');
                    if (args[0] <= 0) throw new CalculationError('Log of non-positive', ErrorCode.InvalidInput);
                    value = args.length === 2 ? Math.log(args[0]) / Math.log(args[1]) : Math.log(args[0]);
                    break;

                case 'sin':
                    if (args.length !== 1) throw new Error('Sin requires 1 argument');
                    value = Math.sin(args[0]);
                    break;

                case 'cos':
                    if (args.length !== 1) throw new Error('Cos requires 1 argument');
                    value = Math.cos(args[0]);
                    break;

                case 'tan':
                    if (args.length !== 1) throw new Error('Tan requires 1 argument');
                    value = Math.tan(args[0]);
                    break;

                case 'abs':
                    if (args.length !== 1) throw new Error('Abs requires 1 argument');
                    value = Math.abs(args[0]);
                    break;

                case 'floor':
                    if (args.length !== 1) throw new Error('Floor requires 1 argument');
                    value = Math.floor(args[0]);
                    break;

                case 'ceil':
                    if (args.length !== 1) throw new Error('Ceil requires 1 argument');
                    value = Math.ceil(args[0]);
                    break;

                // Reduction operations
                case 'sum':
                    value = sum(...args);
                    break;

                case 'product':
                    value = product(...args);
                    break;

                case 'max':
                    if (args.length === 0) throw new Error('Max requires at least 1 argument');
                    value = Math.max(...args);
                    break;

                case 'min':
                    if (args.length === 0) throw new Error('Min requires at least 1 argument');
                    value = Math.min(...args);
                    break;

                case 'avg':
                case 'mean':
                    if (args.length === 0) throw new Error('Average requires at least 1 argument');
                    value = sum(...args) / args.length;
                    break;

                default:
                    throw new CalculationError(`Unknown operation: ${op}`, ErrorCode.InvalidOperation);
            }

            // Round result
            value = this.roundResult(value);
            this._lastResult = value;

            // Create result
            const result: OperationResult = {
                value,
                operation: op,
                inputs: args,
                success: true,
                duration: performance.now() - startTime,
            };

            // Add to history
            this.addToHistory(op, args, value);

            // Notify observers
            this.notify(result);

            return result;

        } catch (error) {
            const result: OperationResult = {
                value: 0,
                operation: op,
                inputs: args,
                success: false,
                error: error instanceof CalculationError
                    ? error
                    : new CalculationError(String(error), ErrorCode.InvalidOperation),
                duration: performance.now() - startTime,
            };

            this.notify(result);
            return result;
        }
    }

    // Implement abstract method
    reset(): void {
        this._memory = 0;
        this._lastResult = 0;
        this._history = [];
    }

    // Private method
    private addToHistory(op: string, args: number[], result: number): void {
        const entry: HistoryEntry = {
            expression: `${op}(${args.join(', ')})`,
            result,
            timestamp: new Date(),
        };

        this._history.push(entry);

        // Trim history if needed
        if (this._history.length > this._historyLimit) {
            this._history = this._history.slice(-this._historyLimit);
        }
    }

    // Memory operations
    get memory(): number {
        return this._memory;
    }

    get lastResult(): number {
        return this._lastResult;
    }

    memoryAdd(value: number): void {
        this._memory += value;
    }

    memorySubtract(value: number): void {
        this._memory -= value;
    }

    memoryClear(): void {
        this._memory = 0;
    }

    memoryRecall(): number {
        return this._memory;
    }

    // Observer pattern implementation
    subscribe(observer: Observer<OperationResult>): Unsubscribe {
        this._observers.push(observer);
        return () => {
            const index = this._observers.indexOf(observer);
            if (index > -1) {
                this._observers.splice(index, 1);
            }
        };
    }

    notify(value: OperationResult): void {
        for (const observer of this._observers) {
            try {
                observer(value);
            } catch (e) {
                console.error('Observer error:', e);
            }
        }
    }
}

/**
 * Scientific calculator extending AdvancedCalculator.
 */
export class ScientificCalculator extends AdvancedCalculator {
    private _angleMode: AngleMode;

    constructor(config: Partial<CalculatorConfig & { angleMode?: AngleMode }> = {}) {
        super(config);
        this._angleMode = config.angleMode ?? 'radians';
    }

    get angleMode(): AngleMode {
        return this._angleMode;
    }

    set angleMode(mode: AngleMode) {
        if (mode !== 'radians' && mode !== 'degrees') {
            throw new Error("Angle mode must be 'radians' or 'degrees'");
        }
        this._angleMode = mode;
    }

    private toRadians(angle: number): number {
        return this._angleMode === 'degrees' ? angle * Math.PI / 180 : angle;
    }

    // Method overloading via overload signatures
    factorial(n: number): number;
    factorial(n: bigint): bigint;
    factorial(n: number | bigint): number | bigint {
        if (typeof n === 'bigint') {
            if (n < 0n) throw new CalculationError('Negative factorial', ErrorCode.NegativeInput);
            let result = 1n;
            for (let i = 2n; i <= n; i++) {
                result *= i;
            }
            return result;
        }

        if (n < 0) throw new CalculationError('Negative factorial', ErrorCode.NegativeInput);
        if (n > 170) throw new CalculationError('Factorial overflow', ErrorCode.Overflow);

        let result = 1;
        for (let i = 2; i <= n; i++) {
            result *= i;
        }
        return result;
    }

    fibonacci(n: number): number {
        if (n < 0) throw new CalculationError('Negative fibonacci', ErrorCode.NegativeInput);
        if (n <= 1) return n;

        let a = 0, b = 1;
        for (let i = 2; i <= n; i++) {
            [a, b] = [b, a + b];
        }
        return b;
    }

    isPrime(n: number): boolean {
        if (n < 2) return false;
        if (n === 2) return true;
        if (n % 2 === 0) return false;

        const sqrt = Math.sqrt(n);
        for (let i = 3; i <= sqrt; i += 2) {
            if (n % i === 0) return false;
        }
        return true;
    }

    primeFactors(n: number): number[] {
        const factors: number[] = [];
        let d = 2;
        while (d * d <= n) {
            while (n % d === 0) {
                factors.push(d);
                n = Math.floor(n / d);
            }
            d++;
        }
        if (n > 1) factors.push(n);
        return factors;
    }

    gcd(a: number, b: number): number {
        a = Math.abs(a);
        b = Math.abs(b);
        while (b !== 0) {
            [a, b] = [b, a % b];
        }
        return a;
    }

    lcm(a: number, b: number): number {
        if (a === 0 || b === 0) return 0;
        return Math.abs(a * b) / this.gcd(a, b);
    }

    // Async calculation
    async calculateAsync(op: string, ...args: number[]): Promise<OperationResult> {
        // Simulate async operation
        await new Promise(resolve => setTimeout(resolve, 1));
        return this.calculate(op, ...args);
    }

    // Generator for Fibonacci sequence
    *fibonacciSequence(limit: number): Generator<number, void, unknown> {
        let a = 0, b = 1;
        for (let i = 0; i < limit; i++) {
            yield a;
            [a, b] = [b, a + b];
        }
    }

    // Async generator for prime numbers
    async *primeGenerator(limit: number): AsyncGenerator<number, void, unknown> {
        let count = 0;
        let n = 2;
        while (count < limit) {
            if (this.isPrime(n)) {
                yield n;
                count++;
            }
            n++;
            // Yield control periodically
            if (n % 100 === 0) {
                await new Promise(resolve => setTimeout(resolve, 0));
            }
        }
    }
}

// Mixin pattern
type Constructor<T = object> = new (...args: unknown[]) => T;

function WithLogging<TBase extends Constructor<BaseCalculator>>(Base: TBase) {
    return class extends Base {
        calculate(op: string, ...args: number[]): OperationResult {
            console.log(`Calculating: ${op}(${args.join(', ')})`);
            const result = super.calculate(op, ...args);
            console.log(`Result: ${result.value}`);
            return result;
        }
    };
}

function WithValidation<TBase extends Constructor<BaseCalculator>>(Base: TBase) {
    return class extends Base {
        calculate(op: string, ...args: number[]): OperationResult {
            for (const arg of args) {
                if (!Number.isFinite(arg)) {
                    return {
                        value: 0,
                        operation: op,
                        inputs: args,
                        success: false,
                        error: new CalculationError('Invalid input', ErrorCode.InvalidInput),
                    };
                }
            }
            return super.calculate(op, ...args);
        }
    };
}

// Create calculator class with mixins
export const LoggingCalculator = WithLogging(AdvancedCalculator);
export const ValidatingCalculator = WithValidation(AdvancedCalculator);
export const LoggingValidatingCalculator = WithLogging(WithValidation(AdvancedCalculator));

// Singleton pattern
export class CalculatorSingleton {
    private static instance: AdvancedCalculator | null = null;

    private constructor() {}

    static getInstance(config?: Partial<CalculatorConfig>): AdvancedCalculator {
        if (!CalculatorSingleton.instance) {
            CalculatorSingleton.instance = new AdvancedCalculator(config);
        }
        return CalculatorSingleton.instance;
    }

    static resetInstance(): void {
        CalculatorSingleton.instance = null;
    }
}
